src/solver/conv_hip_implicit_gemm_fwd_v4r5_xdlops.cpp

```diff
@@ -0,0 +1,1042 @@
+
+namespace miopen {
+namespace solver {
+
+PerformanceImplicitGemmForwardV4R5Xdlops::PerformanceImplicitGemmForwardV4R5Xdlops()
+    : PerformanceImplicitGemmForwardV4R5Xdlops::PerformanceImplicitGemmForwardV4R5Xdlops(
+          4, 4, 1, 4, 4, 1, false, false, 1)
+{
+}
+
+PerformanceImplicitGemmForwardV4R5Xdlops::PerformanceImplicitGemmForwardV4R5Xdlops(
+    int GemmMPerBlock_,
+    int GemmNPerBlock_,
+    int GemmKPerBlock_,
+    int GemmMPerWave_,
+    int GemmNPerWave_,
+    int GemmKPack_,
+    bool GemmAThreadCopyMoreGemmK_,
+    bool GemmBThreadCopyMoreGemmKPack_,
+    int GemmBThreadDataPerRead_GemmN_)
+    : GemmMPerBlock(GemmMPerBlock_),
+      GemmNPerBlock(GemmNPerBlock_),
+      GemmKPerBlock(GemmKPerBlock_),
+      GemmMPerWave(GemmMPerWave_),
+      GemmNPerWave(GemmNPerWave_),
+      GemmKPack(GemmKPack_),
+      GemmAThreadCopyMoreGemmK(GemmAThreadCopyMoreGemmK_),
+      GemmBThreadCopyMoreGemmKPack(GemmBThreadCopyMoreGemmKPack_),
+      GemmBThreadDataPerRead_GemmN(GemmBThreadDataPerRead_GemmN_)
+{
+}
+
+bool PerformanceImplicitGemmForwardV4R5Xdlops::
+operator==(const PerformanceImplicitGemmForwardV4R5Xdlops& other) const
+{
+    // clang-format off
+    return GemmMPerBlock == other.GemmMPerBlock
+        && GemmNPerBlock == other.GemmNPerBlock
+        && GemmKPerBlock == other.GemmKPerBlock
+        && GemmMPerWave == other.GemmMPerWave
+        && GemmNPerWave == other.GemmNPerWave
+        && GemmKPack == other.GemmKPack 
+        && GemmAThreadCopyMoreGemmK  == other.GemmAThreadCopyMoreGemmK
+        && GemmBThreadCopyMoreGemmKPack  == other.GemmBThreadCopyMoreGemmKPack
+        && GemmBThreadDataPerRead_GemmN  == other.GemmBThreadDataPerRead_GemmN;
+    // clang-format on
+}
+
+bool PerformanceImplicitGemmForwardV4R5Xdlops::SetNextValue()
+{
+    do
+    {
+        // list performance parameters in reverse order, in order for tuning to iterate over the
+        // range in normal order
+        if(miopen::IsEnabled(
+               MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_FWD_V4R5_XDLOPS_ADD_VECTOR_LOAD_GEMMN_TUNE_PARAM{}))
+        {
+            if(!NextTwoPower<1, 8>(GemmBThreadDataPerRead_GemmN))
+                break;
+        }
+        if(!NextFlag<false, true>(GemmBThreadCopyMoreGemmKPack))
+            break;
+        if(!NextFlag<false, false>(GemmAThreadCopyMoreGemmK))
+            break;
+        if(!NextTwoPower<1, 8>(GemmKPack))
+            break;
+        if(!NextTwoPower<4, 128>(GemmNPerWave))
+            break;
+        if(!NextTwoPower<4, 128>(GemmMPerWave))
+            break;
+        if(!NextTwoPower<1, 8>(GemmKPerBlock))
+            break;
+        if(!NextTwoPower<4, 256>(GemmNPerBlock))
+            break;
+        if(!NextTwoPower<4, 256>(GemmMPerBlock))
+            break;
+        return false;
+    } while(false);
+
+    return true;
+}
+
+void PerformanceImplicitGemmForwardV4R5Xdlops::EuristicInit(const ConvolutionContext& ctx)
```

Comment:
[Question] Could you please describe the idea of heuristic initialization here? Moving from bigger to smaller values in certain order and then selecting the first "valid" set?

Reply:
Yes. Try from the largest value until find valid values. @asroy correct me if I am wrong.

Reply:
Yes, bigger values corresponds to bigger blockwise GEMM, which give better performance in most cases.

Reply:
[Recommendation] Well, if the iteration order in `SetNextValue()` doesn't matter, then I would recommend:
- Refactoring `SetNextValue()` so it moves from bigger to smaller
- Reusing `SetNextValue()` in `EuristicInit`() thus getting rid of implicit dependency.

Reply:
Good idea